Added initial stub for doc/documentation_contributions.md

jesper-friis opened the pull request.

CasperWA commented on doc/documentation_contributors.md:
````diff
@@ -0,0 +1,58 @@
+* Links become more readable if you place them at the end of the document
+  using square brackets.  Example:
+
+     ```
+     A link to [SOFT].
+     ...
+
+     End of document.
+
+
+     [SOFT]: https://www.sintef.no/en/publications/publication/1553408/
+     ```
+
````
Note, the backslash `\` before the triple backtick should be removed in the actual document.

CasperWA commented on doc/documentation_contributors.md:
````diff
@@ -0,0 +1,58 @@
+     ```
+
+
+The README files are intended to document the overall project or the
+content in a subdirectory.  These should therefore render nicely on GitHub.
+Hence, use [Basic Markdown] or possible [GitHub-flavored Markdown].
````
```suggestion
The README files are intended to document the overall project or the content in a subdirectory.
These should therefore render nicely on GitHub.
Hence, use [Basic Markdown] or possible [GitHub-flavored Markdown].
```

CasperWA replied:
Is `Basic Markdown` a proper name or is it called `CommonMark` as mentioned in the MyST documentation?

CasperWA commented on doc/documentation_contributors.md:
```diff
@@ -0,0 +1,58 @@
+content in a subdirectory.  These should therefore render nicely on GitHub.
+Hence, use [Basic Markdown] or possible [GitHub-flavored Markdown].
+
+The Markdown files in the `doc/` subdirectory are intended to be included
+in the sphinx-generated [online documentation].  Here we can make use of
+the [MyST Markdown extensions], like [admonitions].
```
```suggestion
The Markdown files in the `doc/` subdirectory are intended to be included in the sphinx-generated [online documentation].
Here we can make use of the [MyST Markdown extensions], like [admonitions].
```

CasperWA commented on doc/documentation_contributors.md:
```diff
@@ -0,0 +1,58 @@
+[Markdown]: https://en.wikipedia.org/wiki/Markdown
+[Basic Markdown]: https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet
+[GitHub-flavored Markdown]: https://docs.github.com/en/get-started/writing-on-github
+[MyST Markdown extensions]: https://myst-parser.readthedocs.io/en/latest/syntax/optional.html[online documentation]: https://sintef.github.io/dlite/
```
```suggestion
[MyST Markdown extensions]: https://myst-parser.readthedocs.io/en/latest/syntax/optional.html
[online documentation]: https://sintef.github.io/dlite/
```

CasperWA commented on doc/documentation_contributors.md:
```diff
@@ -0,0 +1,59 @@
+in the sphinx-generated [online documentation].  Here we can make use of
+the [MyST Markdown extensions], like [admonitions].
+
+
+
+
```
```suggestion
```

CasperWA commented on doc/documentation_contributors.md:
```diff
@@ -0,0 +1,116 @@
+Guideline for contributing documentation
+========================================
```
```suggestion
========================================

```
[MD022](https://github.com/DavidAnson/markdownlint/blob/main/doc/md022.md)

CasperWA commented on doc/documentation_contributors.md:
```diff
@@ -0,0 +1,116 @@
+The DLite documentation is written in Markdown.
+This include both the README files and documentation found in the `doc/` subdirectory.
+
+Common to both is that the text should be as easy and natural as possible to read from the terminal.
```
```suggestion
We strive to make the developer experience (DX) as pleasant as possible by ensuring as broad a readability when editing the files as when reading them in a browser.
```

CasperWA commented on doc/documentation_contributors.md:
```diff
@@ -0,0 +1,116 @@
+Common to both is that the text should be as easy and natural as possible to read from the terminal.
+Hence, the following recommendations should be followed:
+
+* Write one sentence per line, in order to get an easier to read output from `git diff`.
```
```suggestion
* Write one sentence per line, in order to get an easier to read output from `git diff`.

  This is essential for the ease of readability and understandability when performing reviews on GitHub.
  As all reviews are performed on GitHub and editing files can be done in multiple different editors, this rule ensures optimal conditions for the GitHub review interface, sacrificing a bit of readability in certain editors.
```

CasperWA commented on doc/documentation_contributors.md:
```diff
@@ -0,0 +1,116 @@
+
+* Write one sentence per line, in order to get an easier to read output from `git diff`.
+
+* For README files, use the underline style format for main and sub-headers.
```
```suggestion
* For README files, use the setext style format for headers.

  As setext styling only supports two header levels, README files cannot include three or more header levels.

```

CasperWA commented on doc/documentation_contributors.md:
```diff
@@ -0,0 +1,116 @@
+* Write one sentence per line, in order to get an easier to read output from `git diff`.
+
+* For README files, use the underline style format for main and sub-headers.
+  For example, start a new page with
+
+      Overview
+      ========
+
+  instead of
+
+      # Overview
```
```suggestion
  The setext header styling:

  ```markdown
  Header level 1
  ==============

  Some text.

  Header level 2
  --------------

  Some text.
  ```
```

CasperWA replied:
Reference to heading style: [MD003](https://github.com/DavidAnson/markdownlint/blob/main/doc/md003.md).

CasperWA commented on doc/documentation_contributors.md:
```diff
@@ -0,0 +1,116 @@
+
+      # Overview
+
+  For documentation the style is up to the writer.
```
I think we should still suggest a best practice here.

CasperWA commented on doc/documentation_contributors.md:
```diff
@@ -0,0 +1,116 @@
+
+  For documentation the style is up to the writer.
+
+* Add two newlines before headers to make them easier to recognise.
+
+  This is especially useful for levels 3-headers and below (which are lines starting with 3 or more hashes (`#`)) to make them easier to recognise as sub-headers.
+
```
```suggestion
```
[MD022](https://github.com/DavidAnson/markdownlint/blob/main/doc/md022.md).
Single surrounding lines are fine. Multiple blank lines is not good practice [MD012](https://github.com/DavidAnson/markdownlint/blob/main/doc/md012.md).

jesper-friis replied:
Removed the bullet point

CasperWA commented on doc/documentation_contributors.md:
```diff
@@ -0,0 +1,116 @@
+
+  This is especially useful for levels 3-headers and below (which are lines starting with 3 or more hashes (`#`)) to make them easier to recognise as sub-headers.
+
+* Avoid the use of html tags.
```
```suggestion
* Avoid the use of raw HTML.
  See [MD033](https://github.com/DavidAnson/markdownlint/blob/main/doc/md033.md).
```

CasperWA commented on doc/documentation_contributors.md:
````diff
@@ -0,0 +1,116 @@
+* Links become more readable if you place them at the end of the document using square brackets.
+  Example:
+
+     ```
+     A link to [SOFT].
+     ...
+
+     End of document.
+
+
+     [SOFT]: https://www.sintef.no/en/publications/publication/1553408/
+     ```
````
```suggestion
   ```markdown
   A link to [SOFT].
   ...

   End of document.

  [SOFT]: https://www.sintef.no/en/publications/publication/1553408/
  ```
```

CasperWA commented on doc/documentation_contributors.md:
```diff
@@ -0,0 +1,116 @@
+
+The README files are intended to document the overall project or the content in a subdirectory.
+These should therefore render nicely on GitHub.
+Hence, use [Basic Markdown] or possible [GitHub-flavored Markdown].
```
```suggestion
Hence, use [CommonMark] or possible [GitHub-flavored Markdown].
```

CasperWA commented on doc/documentation_contributors.md:
```diff
@@ -0,0 +1,116 @@
+Hence, use [Basic Markdown] or possible [GitHub-flavored Markdown].
+Keep the maximum line length at 79 characters so that the source file easily can be viewed in a 80 character wide terminal.
+
+
```
```suggestion
```
[MD012](https://github.com/DavidAnson/markdownlint/blob/main/doc/md012.md)

CasperWA commented on doc/documentation_contributors.md:
```diff
@@ -0,0 +1,116 @@
+The Markdown files in the `doc/` subdirectory are intended to be included in the sphinx-generated [online documentation].
+Here we can make use of the [MyST Markdown extensions], like [admonitions].
+
+
```
```suggestion
```
[MD012](https://github.com/DavidAnson/markdownlint/blob/main/doc/md012.md)

CasperWA commented on doc/documentation_contributors.md:
```diff
@@ -0,0 +1,116 @@
+Keep the maximum line length at 79 characters so that the source file easily can be viewed in a 80 character wide terminal.
+
+
+The Markdown files in the `doc/` subdirectory are intended to be included in the sphinx-generated [online documentation].
```
```suggestion
The Markdown files in the `doc/` subdirectory are intended to be included in the [Sphinx]-generated [online documentation].
```
Include link to Sphinx.

CasperWA commented on doc/documentation_contributors.md:
```diff
@@ -0,0 +1,116 @@
+
+
+The Markdown files in the `doc/` subdirectory are intended to be included in the sphinx-generated [online documentation].
+Here we can make use of the [MyST Markdown extensions], like [admonitions].
```
```suggestion
We make use of the [MyST-Parser extension] to convert all markdown syntax to rich structured text (RST) prior to letting Sphinx generate the HTML (or other) documentation.
```

CasperWA commented on doc/documentation_contributors.md:
```diff
@@ -0,0 +1,116 @@
+
+
+Figures
+-------
+Please place figures in the `doc/figs/` directory.
```
```suggestion
-------

Please place figures in the `doc/figs/` directory.
```
[MD022](https://github.com/DavidAnson/markdownlint/blob/main/doc/md022.md)

CasperWA commented on doc/documentation_contributors.md:
```diff
@@ -0,0 +1,116 @@
+If you use [drawio], it is recommended that you save the figure in svg format.
+Then it renders well in browsers and is easy to find and edit for collaborators.
+
+
```
```suggestion
```
[MD012](https://github.com/DavidAnson/markdownlint/blob/main/doc/md012.md)

CasperWA commented on doc/documentation_contributors.md:
```diff
@@ -0,0 +1,116 @@
+Figures
+-------
+Please place figures in the `doc/figs/` directory.
+If you use [drawio], it is recommended that you save the figure in svg format.
```
```suggestion
If you use [draw.io], it is recommended that you save the figure as vector graphics, e.g., using the SVG format.
```

CasperWA commented on doc/documentation_contributors.md:
```diff
@@ -0,0 +1,116 @@
+-------
+Please place figures in the `doc/figs/` directory.
+If you use [drawio], it is recommended that you save the figure in svg format.
+Then it renders well in browsers and is easy to find and edit for collaborators.
```
```suggestion
This ensures the figures are rendered well with any scaling and are also easy to edit.
```

CasperWA replied:
This was not updated?

CasperWA commented on doc/documentation_contributors.md:
```diff
@@ -0,0 +1,116 @@
+
+
+Examples
+--------
+Write your python examples in the same way as done in the [official Python documentation], by preceeding expressions with `>>> ` and the expected output on the next line below the `>>> `.
```
```suggestion
--------

Write your python examples in the same way as done in the [official Python documentation], by preceeding expressions with `>>> ` and the expected output on the next line below the `>>> `.
```
[MD022](https://github.com/DavidAnson/markdownlint/blob/main/doc/md022.md)

CasperWA replied:
Should these not be respected?

CasperWA commented on doc/documentation_contributors.md:
````diff
@@ -0,0 +1,116 @@
+Write your python examples in the same way as done in the [official Python documentation], by preceeding expressions with `>>> ` and the expected output on the next line below the `>>> `.
+For example
+
+    ```python
+        # Here is a python example
+        >>> x = 1 + 1
+        >>> x
+        2
+
+    ```
````
```suggestion
```markdown
```python
# Here is a python example
>>> x = 1 + 1
>>> x
2
```
```
```

jesper-friis replied:
This crops out the final 3 backticks in the rendered documentation, maning the example incomplete

CasperWA replied:
Did you try this with Sphinx or just the CommonMark/GH rendering?

CasperWA commented on doc/documentation_contributors.md:
````diff
@@ -0,0 +1,116 @@
+        2
+
+    ```
+This allow to check your examples by running
````
```suggestion

This allow to check your examples by running
```
[MD031](https://github.com/DavidAnson/markdownlint/blob/main/doc/md031.md).

CasperWA commented on doc/documentation_contributors.md:
````diff
@@ -0,0 +1,116 @@
+
+    ```
+This allow to check your examples by running
+
+    python -m doctest <filename>
+
````
```suggestion

```shell
python -m doctest <filename>
```

```

jesper-friis replied:
What is the benefit of this? AFAIK, there are no highlighting of shell input, so it only adds unusefull formatting clutter to the documentation source file.

CasperWA replied:
The [shell lexer](https://pygments.org/docs/lexers/#lexers-for-various-shells) is useful if the theme implements formatting for it. As this is up to the theme it is good to always declare the language, should the theme change in the future.

CasperWA commented on doc/documentation_contributors.md:
````diff
@@ -0,0 +1,116 @@
+
+on your markdown file.
+
+It is recommended to indent your code with 4 spaces.
+That makes it easy to copy the correct output into your example.
+For instance, if you write
+
+    ```python
+        >>> 1 + 1  # doctest: +SKIP
+        3
+
+    ```
+
````
```suggestion
```
This should not be an issue and is something that is handled by the Sphinx theme.

jesper-friis replied:
It is, because the `doctest` output is written to the terminal.

CasperWA commented on doc/documentation_contributors.md:
```diff
@@ -0,0 +1,116 @@
+        2
+    **********************************************************************
+
+Now you copy the output following `Got:`, including the 4 indentation spaces,
+into your example.
```
```suggestion
Now you copy the output following `Got:`, including the 4 indentation spaces, into your example.
```

CasperWA commented on doc/documentation_contributors.md:
```diff
@@ -0,0 +1,116 @@
+
+on your markdown file.
+
+It is recommended to indent your code with 4 spaces.
```
All text from here and down does not belong in this document in my opinion.
These are usage examples and description of how to use `doctest`. Simply refer to the `doctest` documentation instead or write a separate documentation about how to test documentation.

jesper-friis replied:
Ok, moved the text into a separate file.

CasperWA commented on doc/documentation_contributors.md:
```diff
@@ -0,0 +1,116 @@
+
+Note that the comment `#doctest: +SKIP` was added to the above example in order to not triggering a doctest failure when validating the markdown file containing this guideline.
+
+
+
```
```suggestion
```
[MD012](https://github.com/DavidAnson/markdownlint/blob/main/doc/md012.md)

CasperWA commented:
There are several suggestions from my previous review you have not addressed it seems: 
![image](https://user-images.githubusercontent.com/43357585/210565490-c07ed84c-12c5-4ea0-9c4b-f396a114dae9.png)

CasperWA commented:
> There are several suggestions from my previous review you have not addressed it seems (...)

Please click "Load more" above:
![image](https://user-images.githubusercontent.com/43357585/210573320-e3544d28-ce32-40e8-87c2-0b045de8f130.png)

jesper-friis commented:
> There are several suggestions from my previous review you have not addressed it seems: ![image](https://user-images.githubusercontent.com/43357585/210565490-c07ed84c-12c5-4ea0-9c4b-f396a114dae9.png)

Ok. Updated the first and removed the second point you commented on

CasperWA commented on doc/documentation_contributors.md:
```diff
@@ -0,0 +1,93 @@
+
+The README files are intended to document the overall project or the content in a subdirectory.
+These should therefore render nicely on GitHub.
+Hence, use [CommonMark] or possible [GitHub-flavored Markdown].
```
```suggestion
Hence, use [CommonMark] or [GitHub-flavored Markdown].
```

jesper-friis replied:
CommonMark is preferred, since we it is not guaranteed that we will be using github forever.

CasperWA commented on doc/documentation_contributors.md:
```diff
@@ -0,0 +1,93 @@
+The README files are intended to document the overall project or the content in a subdirectory.
+These should therefore render nicely on GitHub.
+Hence, use [CommonMark] or possible [GitHub-flavored Markdown].
+For README files, keep the maximum line length at 79 characters so that the source file easily can be viewed in a 80 character wide terminal.
```
```suggestion
For README files, keep the maximum line length at 79 characters for easy viewing in a 80 character wide terminal.
```

CasperWA commented on doc/documentation_contributors.md:
```diff
@@ -0,0 +1,93 @@
+Hence, use [CommonMark] or possible [GitHub-flavored Markdown].
+For README files, keep the maximum line length at 79 characters so that the source file easily can be viewed in a 80 character wide terminal.
+
+The Markdown files in the `doc/` subdirectory are intended to be included in the [sphinx]-generated [online documentation].
```
```suggestion
The Markdown files in the `doc/` subdirectory are intended to be included in the [Sphinx]-generated [online documentation].
```

CasperWA commented on doc/documentation_contributors.md:
```diff
@@ -0,0 +1,93 @@
+
+on your markdown file.
+
+It is recommended to indent your code with 4 spaces.
+That makes it easy to copy the correct doctest output into your example.
+For more info, see the guide for [documentation testing].
```
This makes absolutely no sense to me, sorry. Why is it necessary/convenient to indent the code block with 4 spaces as the first thing? When copying the code, it should be complete by itself, it will _not_ be complete by itself if everything is indented?
You're even indenting the `>>>`, which is part of the shell even. It has nothing to do with the actual content one is writing...

CasperWA commented on doc/documentation_contributors.md:
```diff
@@ -0,0 +1,93 @@
+[CommonMark]: https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet
+[GitHub-flavored Markdown]: https://docs.github.com/en/get-started/writing-on-github
+[MyST Markdown extensions]: https://myst-parser.readthedocs.io/en/latest/syntax/optional.html
+[sphinx]: https://www.sphinx-doc.org/
```
```suggestion
[Sphinx]: https://www.sphinx-doc.org/
```

CasperWA commented on doc/documentation_testing.md:
```diff
@@ -0,0 +1,38 @@
+Testing Python examples in the documentation
+============================================
+We can use the Python doctest module for testing examples in the documentation.
```
```suggestion
============================================

We can use the Python [`doctest`](https://docs.python.org/3/library/doctest.html) module for testing examples in the documentation.
```
[MD022](https://github.com/DavidAnson/markdownlint/blob/main/doc/md022.md)

CasperWA commented on doc/documentation_testing.md:
```diff
@@ -0,0 +1,38 @@
+============================================
+We can use the Python doctest module for testing examples in the documentation.
+Run it with
+
+    python -m doctest <filename>
+
```
```suggestion

```shell
python -m doctest <filename>
```

```

CasperWA commented on doc/documentation_testing.md:
```diff
@@ -0,0 +1,38 @@
+
+Note that the comment `#doctest: +SKIP` was added to the above example in order to not triggering a doctest failure when validating the markdown file containing this guideline.
+
+
+
```
```suggestion
```
[MD012](https://github.com/DavidAnson/markdownlint/blob/main/doc/md012.md)

CasperWA commented on doc/documentation_testing.md:
````diff
@@ -0,0 +1,38 @@
+
+For instance, assume that you have the following example in your documentation
+
+    ```python
+        >>> 1 + 1  # doctest: +SKIP
+        3
+
+    ```
````
```suggestion
```python
>>> 1 + 1  # doctest: +SKIP
3
```
```

jesper-friis replied:
Doesn't work. The closing inner backticks are stripped off...

CasperWA commented on doc/documentation_testing.md:
```diff
@@ -0,0 +1,38 @@
+
+`python -m doctest documentation_testing.md` would print the following message:
+
+    **********************************************************************
+    File "documentation_testing.md", line 13, in documentation_testing.md
+    Failed example:
+        1 + 1
+    Expected:
+        3
+    Got:
+        2
+    **********************************************************************
```
```suggestion
```shell
**********************************************************************
File "documentation_testing.md", line 13, in documentation_testing.md
Failed example:
    1 + 1
Expected:
    3
Got:
    2
**********************************************************************
```
```

CasperWA commented on doc/documentation_testing.md:
```diff
@@ -0,0 +1,38 @@
+
+Now you copy the output following `Got:`, including the 4 indentation spaces, into your example.
+
+Note that the comment `#doctest: +SKIP` was added to the above example in order to not triggering a doctest failure when validating the markdown file containing this guideline.
```
```suggestion
Note that the comment `#doctest: +SKIP` was added to the above example in order to not trigger a doctest failure when validating the markdown file containing this guideline.
```

CasperWA commented:
I will apply the "easy" suggested changes (`sphinx` -> `Sphinx`, adding an external reference), but leave the "controversial" suggestions. Then merge this as it feels pointless to keep this under review. I'd rather open a new PR with my more specific suggested changes.

CasperWA reviewed (approved):
Will use the currently unresolved suggestions in a separate PR.
